plugin/pkg/shoot/validator/admission.go

```diff
@@ -628,6 +634,107 @@ func (c *validationContext) validateShootHibernation(a admission.Attributes) err
+			credentialsBinding.CredentialsRef.APIVersion != corev1.SchemeGroupVersion.String() ||
+			credentialsBinding.CredentialsRef.Name != secretBinding.SecretRef.Name ||
+			credentialsBinding.CredentialsRef.Namespace != secretBinding.SecretRef.Namespace {
+			return admission.NewForbidden(a, errors.New("it is not allowed to change the referenced Secret when migrating from SecretBindingName to CredentialsBindingName"))
+		}
+	}
+
+	return nil
+}
+
+func (c *validationContext) validateCredentialsBindingChange(
+	ctx context.Context,
+	a admission.Attributes,
+	auth authorizer.Authorizer,
+	credentialsBindingLister securityv1alpha1listers.CredentialsBindingLister,
+) error {
+	getAttributesRecord := func(credentialsBinding *securityv1alpha1.CredentialsBinding) (authorizer.AttributesRecord, error) {
+		var (
+			credentialsAPIGroup   string
+			credentialsAPIVersion string
+			credentialsResource   string
+		)
+		if credentialsBinding.CredentialsRef.APIVersion == corev1.SchemeGroupVersion.String() {
+			credentialsAPIGroup = corev1.SchemeGroupVersion.Group
+			credentialsAPIVersion = corev1.SchemeGroupVersion.Version
+			credentialsResource = "secrets"
+		} else if credentialsBinding.CredentialsRef.APIVersion == securityv1alpha1.SchemeGroupVersion.String() {
+			credentialsAPIGroup = securityv1alpha1.SchemeGroupVersion.Group
+			credentialsAPIVersion = securityv1alpha1.SchemeGroupVersion.Version
+			credentialsResource = "workloadidentities"
+		} else {
+			return authorizer.AttributesRecord{}, errors.New("unknown credentials ref: CredentialsBinding is referencing neither a Secret nor a WorkloadIdentity")
+		}
+		return authorizer.AttributesRecord{
+			User:            a.GetUserInfo(),
+			Verb:            "get",
+			APIGroup:        credentialsAPIGroup,
+			APIVersion:      credentialsAPIVersion,
+			Resource:        credentialsResource,
+			Namespace:       credentialsBinding.CredentialsRef.Namespace,
+			Name:            credentialsBinding.CredentialsRef.Name,
+			ResourceRequest: true,
+		}, nil
+	}
+
+	// Prevent users from changing the credentials binding unless they have read permissions for both old and new credentials.
+	// This ensures that if a user has access to a cluster
+	if c.oldShoot.Spec.CredentialsBindingName != nil && c.shoot.Spec.CredentialsBindingName != nil &&
+		*c.oldShoot.Spec.CredentialsBindingName != *c.shoot.Spec.CredentialsBindingName {
+		oldCredentialsBinding, err := credentialsBindingLister.CredentialsBindings(c.oldShoot.Namespace).Get(*c.oldShoot.Spec.CredentialsBindingName)
+		if err != nil {
+			return apierrors.NewInternalError(fmt.Errorf("could not retrieve previously referenced credentials binding: %+v", err.Error()))
+		}
+
+		oldCredentialsBindingAttributesRecord, err := getAttributesRecord(oldCredentialsBinding)
+		if err != nil {
+			return admission.NewForbidden(a, err)
+		}
+
+		if decision, _, _ := auth.Authorize(ctx, oldCredentialsBindingAttributesRecord); decision != authorizer.DecisionAllow {
```

Comment:
Shouldn't we also check for any errors returned by `Authorize` and report them for more clarity instead of only relying on whether the decision was allowed or not?

Reply:
I was following the convention used here: https://github.com/gardener/gardener/blob/b40ab68dc8c06b97e075cdf084709ac8d1196245/plugin/pkg/global/resourcereferencemanager/admission.go#L945

But maybe it makes sense to include the error in the message if there is such. WDYT?

Reply:
There's a check for a similar authorization call in the `shoot/validator/admission.go` file: https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/validator/admission.go#L557-L570
But I would also wrap the error with additional text to indicate the context in which it originates.